File: src/VitalSynth.Core/SlowSignals.cs
using System.Diagnostics;

namespace VitalSynth.Core;

/// <summary>
/// Slow-moving signals: respiration LFO, SpO2 drift, temperature drift.
/// </summary>
public sealed class SlowSignals(Cfg cfg)
{
    private readonly Random _rng = new(0x5B02);
    private readonly Stopwatch _clock = Stopwatch.StartNew();
    private double _respPhase;
    private double _spo2Walk;
    private double _tempWalk;

    public double LastSpo2 { get; private set; } = cfg.Spo2Base;
    public double LastTempC { get; private set; } = cfg.TempBase;

    public double Respiration(double dt)
    {
        _respPhase += 2 * Math.PI * cfg.RespRpm / 60 * dt;
        return Math.Sin(_respPhase);
    }

    public double AdvanceSpo2(double dt)
    {
        _spo2Walk = Math.Clamp(_spo2Walk + (_rng.NextDouble() - 0.5) * cfg.Spo2Drift * dt, -1.5, 1.5);
        var value = cfg.Spo2Base + _spo2Walk;
        if (cfg.Scenario is Scenario.Desaturation)
            value -= 8 * (1 - Math.Exp(-_clock.Elapsed.TotalSeconds / 15));
        LastSpo2 = Math.Clamp(value, 70, 100);
        return LastSpo2;
    }

    public double AdvanceTemperature(double dt)
    {
        _tempWalk += (_rng.NextDouble() - 0.5) * cfg.TempDrift * dt;
        LastTempC = cfg.TempBase + _tempWalk;
        return LastTempC;
    }
}

File: src/VitalSynth.Core.FS/SlowSignals.fs
namespace VitalSynth.Core.FS

open System
open System.Diagnostics

/// Slow-moving signals: respiration LFO, SpO2 drift, temperature drift.
///
/// Boundary types carry units of measure; internal scratch variables
/// (drift walks etc.) stay raw <c>float</c> — wrapping every line would
/// only add noise. This is the same pragmatic split used in the C# version.
type SlowSignals(cfg: Cfg) =
    let rng = Random(0x5B02)
    let clock = Stopwatch.StartNew()
    let mutable respPhase = 0.0
    let mutable spo2Walk = 0.0
    let mutable tempWalk = 0.0
    let mutable lastSpo2 = cfg.Spo2Base
    let mutable lastTemp = cfg.TempBase

    member _.LastSpo2 = lastSpo2
    member _.LastTemperature = lastTemp

    member _.Respiration (dt: float<s>) : float =
        respPhase <- respPhase + 2.0 * Math.PI * float cfg.RespiratoryRate / 60.0 * float dt
        sin respPhase

    member _.AdvanceSpo2 (dt: float<s>) : float<pct> =
        spo2Walk <-
            Math.Clamp(
                spo2Walk + (rng.NextDouble() - 0.5) * cfg.Spo2Drift * float dt,
                -1.5, 1.5)
        let mutable value = float cfg.Spo2Base + spo2Walk
        match cfg.Scenario with
        | Desaturation _ ->
            value <- value - 8.0 * (1.0 - Math.Exp(-clock.Elapsed.TotalSeconds / 15.0))
        | Normal | Tachycardia | Bradycardia | AFib _ | Asystole ->
            ()
        lastSpo2 <- Math.Clamp(value, 70.0, 100.0) * 1.0<pct>
        lastSpo2

    member _.AdvanceTemperature (dt: float<s>) : float<degC> =
        tempWalk <- tempWalk + (rng.NextDouble() - 0.5) * cfg.TempDrift * float dt
        lastTemp <- (float cfg.TempBase + tempWalk) * 1.0<degC>
        lastTemp
